File: ts/src/metadata.ts
import { xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { BinaryReader, BinaryWriter } from "./binary.js";

// MetadataMode controls how Metadata records are handled during encryption.
export type MetadataMode = "plaintext" | "encrypt" | "encrypt-all";

const NONCE_SIZE = 24;
const FLAG_ENCRYPT = 0x00;     // name plaintext, map encrypted
const FLAG_ENCRYPT_ALL = 0x01; // name + map both encrypted

export interface EncryptedMetadata {
  flags: number;
  name: string;   // plaintext name (empty when flags = FLAG_ENCRYPT_ALL)
  nonce: Uint8Array;
  encryptedData: Uint8Array;
}

// metadataAAD builds AEAD additional data for one encrypted metadata record.
// flags=0x00: AAD = fileId + uint32LE(len(name)) + nameBytes
// flags=0x01: AAD = fileId only
function metadataAAD(fileId: Uint8Array, flags: number, name: string): Uint8Array {
  const w = new BinaryWriter();
  w.writeBytes(fileId);
  if (flags === FLAG_ENCRYPT) {
    w.writeString(name); // 4-byte LE prefix + utf8
  }
  return w.toUint8Array();
}

export function encodeEncryptedMetadata(m: EncryptedMetadata): Uint8Array {
  const w = new BinaryWriter();
  w.writeUint8(m.flags);
  w.writeString(m.name);
  w.writePrefixedBytes(m.nonce);
  w.writePrefixedBytes(m.encryptedData);
  return w.toUint8Array();
}

export function decodeEncryptedMetadata(data: Uint8Array): EncryptedMetadata {
  const r = new BinaryReader(data);
  if (r.remaining < 1) throw new Error("encrypted metadata record too short");
  const flags = r.readUint8();
  if (flags !== FLAG_ENCRYPT && flags !== FLAG_ENCRYPT_ALL) {
    throw new Error(`unknown encrypted metadata flags 0x${flags.toString(16).padStart(2, "0")}`);
  }
  const name = r.readString();
  const nonce = new Uint8Array(r.readPrefixedBytes());
  const encryptedData = new Uint8Array(r.readPrefixedBytes());
  return { flags, name, nonce, encryptedData };
}

// encryptMetadata encrypts one raw MCAP Metadata record payload according to mode.
// metadataPayload is the raw bytes of the Metadata record (name + map).
export function encryptMetadata(
  metadataPayload: Uint8Array,
  symKey: Uint8Array,
  fileId: Uint8Array,
  mode: MetadataMode,
): EncryptedMetadata {
  const nonce = new Uint8Array(NONCE_SIZE);
  crypto.getRandomValues(nonce);

  if (mode === "encrypt") {
    // Extract name from the payload (uint32 LE prefix + utf8 bytes).
    const r = new BinaryReader(metadataPayload);
    const name = r.readString(); // advances r.offset past the name field
    // Map bytes = everything after the name field.
    const mapBytes = metadataPayload.subarray(r.offset);
    const aad = metadataAAD(fileId, FLAG_ENCRYPT, name);
    const encryptedData = xchacha20poly1305(symKey, nonce, aad).encrypt(mapBytes);
    return { flags: FLAG_ENCRYPT, name, nonce, encryptedData };
  }

  // encrypt-all: encrypt the full payload
  const aad = metadataAAD(fileId, FLAG_ENCRYPT_ALL, "");
  const encryptedData = xchacha20poly1305(symKey, nonce, aad).encrypt(metadataPayload);
  return { flags: FLAG_ENCRYPT_ALL, name: "", nonce, encryptedData };
}

// decryptMetadata decrypts one EncryptedMetadata and returns the plaintext
// MCAP Metadata record payload (name + map bytes).
export function decryptMetadata(
  em: EncryptedMetadata,
  symKey: Uint8Array,
  fileId: Uint8Array,
): Uint8Array {
  if (em.nonce.length !== NONCE_SIZE) {
    throw new Error(`encrypted metadata: nonce length ${em.nonce.length} invalid (want ${NONCE_SIZE})`);
  }
  if (em.encryptedData.length < 16) {
    throw new Error(`encrypted metadata: ciphertext too short (${em.encryptedData.length} bytes, minimum 16)`);
  }
  const aad = metadataAAD(fileId, em.flags, em.name);
  let plain: Uint8Array;
  try {
    plain = xchacha20poly1305(symKey, em.nonce, aad).decrypt(em.encryptedData);
  } catch {
    throw new Error("decrypt metadata: AEAD authentication failed");
  }

  if (em.flags === FLAG_ENCRYPT_ALL) {
    return plain; // full metadata payload
  }

  // flags === FLAG_ENCRYPT: plain is map bytes only; prepend the plaintext name
  const nameBytes = new TextEncoder().encode(em.name);
  const out = new Uint8Array(4 + nameBytes.length + plain.length);
  new DataView(out.buffer).setUint32(0, nameBytes.length, true);
  out.set(nameBytes, 4);
  out.set(plain, 4 + nameBytes.length);
  return out;
}

File: pkg/mcapencrypt/metadata.go
package mcapencrypt

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"

	"golang.org/x/crypto/chacha20poly1305"
)

// MetadataMode controls how Metadata records are handled during encryption.
type MetadataMode string

const (
	// MetadataPlaintext passes Metadata records through unchanged (default).
	// Compatible with all MCAP readers.
	MetadataPlaintext MetadataMode = "plaintext"

	// MetadataEncrypt encrypts the metadata map while keeping the record name
	// readable without a key. Stored as EncryptedMetadata (opcode 0x83).
	MetadataEncrypt MetadataMode = "encrypt"

	// MetadataEncryptAll encrypts both the name and the map.
	// No information is visible without the private key.
	// Stored as EncryptedMetadata (opcode 0x83) with an empty plaintext name.
	MetadataEncryptAll MetadataMode = "encrypt-all"
)

// metadataFlag values written into the EncryptedMetadata flags byte.
const (
	metadataFlagEncrypt    = byte(0x00) // name plaintext, map encrypted
	metadataFlagEncryptAll = byte(0x01) // name + map both encrypted
)

// EncryptedMetadata is the on-disk format for an encrypted MCAP metadata
// record (opcode 0x83).
//
// Layout:
//
//	flags          uint8   — 0x00: name plaintext / 0x01: name+map encrypted
//	name           string  — plaintext name (4-byte LE length + utf8; empty when flags=0x01)
//	nonce          bytes   — 24-byte XChaCha20Poly1305 nonce (4-byte LE length prefix)
//	encrypted_data bytes   — ciphertext (4-byte LE length prefix)
//
// flags=0x00: encrypted_data covers only the map bytes; AAD = file_id + uint32(len(name)) + name.
// flags=0x01: encrypted_data covers the full metadata payload (name+map); AAD = file_id.
type EncryptedMetadata struct {
	Flags         byte
	Name          string
	Nonce         []byte
	EncryptedData []byte
}

func (m *EncryptedMetadata) Encode() []byte {
	name := []byte(m.Name)

	n := 1 +
		4 + len(name) +
		4 + len(m.Nonce) +
		4 + len(m.EncryptedData)

	buf := make([]byte, n)
	o := 0

	buf[o] = m.Flags
	o++
	put32 := func(v uint32) { binary.LittleEndian.PutUint32(buf[o:], v); o += 4 }
	putBytes := func(b []byte) { put32(uint32(len(b))); copy(buf[o:], b); o += len(b) }

	putBytes(name)
	putBytes(m.Nonce)
	putBytes(m.EncryptedData)

	return buf
}

// DecodeEncryptedMetadata parses a raw EncryptedMetadata record payload.
func DecodeEncryptedMetadata(data []byte) (*EncryptedMetadata, error) {
	if len(data) < 1 {
		return nil, fmt.Errorf("encrypted metadata record too short (%d bytes)", len(data))
	}
	m := &EncryptedMetadata{}
	o := 0

	m.Flags = data[o]
	o++

	if m.Flags != metadataFlagEncrypt && m.Flags != metadataFlagEncryptAll {
		return nil, fmt.Errorf("unknown encrypted metadata flags 0x%02x", m.Flags)
	}

	getBytes := func() ([]byte, error) {
		if o+4 > len(data) {
			return nil, fmt.Errorf("truncated at offset %d reading length", o)
		}
		n := int(binary.LittleEndian.Uint32(data[o:]))
		o += 4
		if o+n > len(data) {
			return nil, fmt.Errorf("truncated at offset %d reading %d bytes", o, n)
		}
		v := make([]byte, n)
		copy(v, data[o:o+n])
		o += n
		return v, nil
	}
	getString := func() (string, error) {
		b, err := getBytes()
		return string(b), err
	}

	var err error
	if m.Name, err = getString(); err != nil {
		return nil, fmt.Errorf("read name: %w", err)
	}
	if m.Nonce, err = getBytes(); err != nil {
		return nil, fmt.Errorf("read nonce: %w", err)
	}
	if m.EncryptedData, err = getBytes(); err != nil {
		return nil, fmt.Errorf("read encrypted_data: %w", err)
	}
	return m, nil
}

// metadataAAD builds AEAD additional data for one encrypted metadata record.
//
// flags=0x00: AAD = file_id + uint32(len(name)) + name
// flags=0x01: AAD = file_id only
func metadataAAD(fileID []byte, flags byte, name string) []byte {
	if flags == metadataFlagEncryptAll {
		buf := make([]byte, fileIDSize)
		copy(buf, fileID)
		return buf
	}
	// flags == metadataFlagEncrypt: bind plaintext name to ciphertext
	nb := []byte(name)
	buf := make([]byte, 0, fileIDSize+4+len(nb))
	buf = append(buf, fileID...)
	var n [4]byte
	binary.LittleEndian.PutUint32(n[:], uint32(len(nb)))
	buf = append(buf, n[:]...)
	buf = append(buf, nb...)
	return buf
}

// splitMetadataPayload splits a raw MCAP Metadata payload into the name string
// and the remaining map bytes (everything after the name field).
func splitMetadataPayload(data []byte) (name string, mapBytes []byte, err error) {
	if len(data) < 4 {
		return "", nil, fmt.Errorf("metadata payload too short (%d bytes)", len(data))
	}
	n := int(binary.LittleEndian.Uint32(data))
	if 4+n > len(data) {
		return "", nil, fmt.Errorf("truncated metadata name (need %d bytes, have %d)", 4+n, len(data))
	}
	return string(data[4 : 4+n]), data[4+n:], nil
}

// reassembleMetadataPayload rebuilds a full MCAP Metadata payload from a
// plaintext name and decrypted map bytes.
func reassembleMetadataPayload(name string, mapBytes []byte) []byte {
	nb := []byte(name)
	buf := make([]byte, 4+len(nb)+len(mapBytes))
	binary.LittleEndian.PutUint32(buf, uint32(len(nb)))
	copy(buf[4:], nb)
	copy(buf[4+len(nb):], mapBytes)
	return buf
}

// encryptMetadataRecord encrypts one raw MCAP Metadata payload according to mode.
// It returns an EncryptedMetadata ready to encode, or an error.
func encryptMetadataRecord(payload, symKey, fileID []byte, mode MetadataMode) (*EncryptedMetadata, error) {
	aead, err := chacha20poly1305.NewX(symKey)
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, chacha20poly1305.NonceSizeX)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}

	switch mode {
	case MetadataEncrypt:
		name, mapBytes, err := splitMetadataPayload(payload)
		if err != nil {
			return nil, fmt.Errorf("split metadata payload: %w", err)
		}
		aad := metadataAAD(fileID, metadataFlagEncrypt, name)
		return &EncryptedMetadata{
			Flags:         metadataFlagEncrypt,
			Name:          name,
			Nonce:         nonce,
			EncryptedData: aead.Seal(nil, nonce, mapBytes, aad),
		}, nil

	case MetadataEncryptAll:
		aad := metadataAAD(fileID, metadataFlagEncryptAll, "")
		return &EncryptedMetadata{
			Flags:         metadataFlagEncryptAll,
			Name:          "",
			Nonce:         nonce,
			EncryptedData: aead.Seal(nil, nonce, payload, aad),
		}, nil

	default:
		return nil, fmt.Errorf("unsupported metadata mode %q", mode)
	}
}

// decryptMetadataRecord decrypts one EncryptedMetadata and returns the
// plaintext MCAP Metadata payload (name + map bytes).
func decryptMetadataRecord(em *EncryptedMetadata, symKey, fileID []byte) ([]byte, error) {
	if len(em.Nonce) != chacha20poly1305.NonceSizeX {
		return nil, fmt.Errorf("nonce length %d invalid (want %d)", len(em.Nonce), chacha20poly1305.NonceSizeX)
	}
	if len(em.EncryptedData) < 16 {
		return nil, fmt.Errorf("ciphertext too short (%d bytes, minimum 16)", len(em.EncryptedData))
	}
	aead, err := chacha20poly1305.NewX(symKey)
	if err != nil {
		return nil, err
	}
	aad := metadataAAD(fileID, em.Flags, em.Name)
	plain, err := aead.Open(nil, em.Nonce, em.EncryptedData, aad)
	if err != nil {
		return nil, fmt.Errorf("AEAD authentication failed")
	}

	switch em.Flags {
	case metadataFlagEncryptAll:
		// plain is the full metadata payload
		return plain, nil
	case metadataFlagEncrypt:
		// plain is map bytes only; prepend the plaintext name
		return reassembleMetadataPayload(em.Name, plain), nil
	default:
		return nil, fmt.Errorf("unknown flags 0x%02x", em.Flags)
	}
}
